text/0091-dht-record-creation-time.md

```diff
@@ -0,0 +1,108 @@
+| **Authors**     | Alex Gheorghe (alexggh)                                                                     |
+
+## Summary
+
+Extend the DHT authority discovery records with a signed creation time, so that nodes can determine which record is newer and always decide to prefer the newer records to the old ones.
+
+## Motivation
+
+Currently, we use the Kademlia DHT for storing records regarding the p2p address of an authority discovery key, the problem is that if the nodes decide to change its PeerId/Network key it will publish a new record, however because of the distributed and replicated nature of the DHT there is no way to tell which record is newer so both old PeerId and the new PeerId will live in the network until the old one expires(36h), that creates all sort of problem and leads to the node changing its address not being properly connected for up to 36h. 
+
+After this RFC, nodes are extended to decide to keep the new record and propagate the new record to nodes that have the old record stored, so in the end all the nodes will converge faster to the new record(in the order of minutes, not 36h)
+
+Implementation of the rfc: https://github.com/paritytech/polkadot-sdk/pull/3786.
+
+Current issue without this enhacement: https://github.com/paritytech/polkadot-sdk/issues/3673 
+
+## Stakeholders
+
+Polkadot node developers.
+
+## Explanation
+
+This RFC heavily relies on the functionalities of the Kademlia DHT already in use by Polkadot.
+You can find a link to the specification [here](https://github.com/libp2p/specs/tree/master/kad-dht).
+
+In a nutshell, on a specific node the current authority-discovery protocol publishes Kademila DHT records at startup and periodically. The records contain the full address of the node for each authorithy key it owns. The node tries also to find the full address of all authorities in the network by querying the DHT and picking up the first record it finds for each of the authority id it found on chain.
+
+The authority discovery DHT records use the protobuf protocol and the current format is specified [here](https://github.com/paritytech/polkadot-sdk/blob/313fe0f9a277f27a4228634f0fb15a1c3fa21271/substrate/client/authority-discovery/src/worker/schema/dht-v2.proto#L4). This RFC proposese extending the schema in a backwards compatible manner by adding a new optional `creation_time` field to `SignedAuthorityRecord` which nodes can use to determine which of the record is newer.
+
+Diff of `dht-v3.proto` vs `dht-v2.proto`
+
+```
+@@ -1,10 +1,10 @@
+ syntax = "proto3";
+
+-package authority_discovery_v2;
++package authority_discovery_v3;
+
+@@ -13,11 +13,21 @@
+ 	bytes public_key = 2;
+ }
+
++// Information regarding the creation data of the record
++message TimestampInfo {
++	// Time since UNIX_EPOCH in nanoseconds, scale encoded
++	bytes timestamp = 1;
++	// A signature of the creation time.
++	bytes signature = 2;
```

Comment:
If you the timestamp to the content of `record` itself (which is also a protobuf struct), it would avoid adding yet another signature.

Reply:
I added a new one for backwards compatibility, but I think you are right I could get by just adding the creation_time field to AuthorityRecord, old nodes seems do the signature checks before decoding it.

Reply:
Done.